File: 518_Coin Change II.java
// 时间复杂度：O(n⋅amount)，其中 n 为 coins 的长度。
// 空间复杂度：O(amount)。

class Solution {
    public int change(int amount, int[] coins) {
        // 完全背包
        int[] dp = new int[amount + 1];
        dp[0] = 1;

        for (int c : coins) {
            for (int i = c; i <= amount; i++) {
                dp[i] += dp[i - c];
            }
        }
        return dp[amount];
    }
}

File: 518_Coin Change II.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        bag = [1] + [0] * amount

        for c in coins:
            for i in range(c, amount + 1):
                if bag[i-c] != 0: bag[i] += bag[i-c]
        return bag[-1]